src/conn_poll.c

```diff
@@ -141,299 +313,487 @@ int conn_poll_prepare(conn_registry_t *registry, pfds_record_t *pfds, timestamp_
+		connection->send_size = pending;
+		connection->send_offset = 0;
+	} else if (pending == 0) {
+		connection->send_size = 0;
+		connection->send_offset = 0;
 	}
 
-	if (pfd->revents & POLLERR || (pfd->revents & POLLHUP && !(pfd->revents & POLLIN))) {
-		JLOG_DEBUG("TCP connection failed");
-		conn_poll_change_tcp_fail(agent);
-		return;
+	size_t frame_size = size + 2;
+	size_t required = connection->send_size + frame_size;
+	if (required > connection->send_capacity) {
+		size_t new_capacity = connection->send_capacity ? connection->send_capacity : 512;
+		while (new_capacity < required)
+			new_capacity *= 2;
+		uint8_t *new_buffer = realloc(connection->send_buffer, new_capacity);
+		if (!new_buffer) {
+			JLOG_FATAL("Memory allocation failed for TCP send buffer");
+			return -1;
+		}
+		connection->send_buffer = new_buffer;
+		connection->send_capacity = new_capacity;
 	}
 
-	if (pfd->revents & POLLOUT) {
-		if (conn_impl->tcp_state == TCP_STATE_CONNECTING) {
-			int err = 0;
-			socklen_t errlen = sizeof(err);
-			if (getsockopt(conn_impl->tcp_sock, SOL_SOCKET, SO_ERROR, (char *)&err, &errlen) != 0) {
-				JLOG_DEBUG("Failed to get socket error code");
-				conn_poll_change_tcp_fail(agent);
-				return;
-			}
+	uint8_t *dst = connection->send_buffer + connection->send_size;
+	dst[0] = (uint8_t)((size >> 8) & 0xFF);
+	dst[1] = (uint8_t)(size & 0xFF);
+	memcpy(dst + 2, data, size);
+	connection->send_size += frame_size;
 
-			if (err != 0) {
-				JLOG_DEBUG("TCP connection failed, errno=%d", err);
-				conn_poll_change_tcp_fail(agent);
-				return;
-			}
+	return conn_poll_tcp_flush(connection);
+}
 
-			conn_poll_change_tcp_state(agent, TCP_STATE_CONNECTED);
-		} else {
-			tcp_ice_write_context_t *context = &conn_impl->tcp_ice_write_context;
-			if(context->pending) {
-				int ret = tcp_ice_write(conn_impl->tcp_sock, NULL, 0, context);
-				if (ret >= 0) {
-					JLOG_DEBUG("Finished sending ICE-TCP datagram");
-				} else if (ret == -SEAGAIN || ret == -SEWOULDBLOCK) {
-					JLOG_WARN("TCP send failed, errno=%d", -ret);
-					conn_poll_change_tcp_fail(agent);
-					return;
-				}
-			}
+static int conn_poll_accept_tcp_connections(juice_agent_t *agent, conn_impl_t *conn_impl) {
+	if (conn_impl->tcp_listener == INVALID_SOCKET)
+		return 0;
+
+	for (;;) {
+		addr_record_t client;
+		client.len = sizeof(client.addr);
+		socket_t client_sock = accept(conn_impl->tcp_listener, (struct sockaddr *)&client.addr,
+		                            &client.len);
+		if (client_sock == INVALID_SOCKET) {
+			if (sockerrno == SEAGAIN || sockerrno == SEWOULDBLOCK)
+				return 0;
+			if (sockerrno == SEINTR)
+				continue;
+			JLOG_WARN("TCP accept failed, errno=%d", sockerrno);
+			return -1;
+		}
+		client.socktype = SOCK_STREAM;
+		addr_unmap_inet6_v4mapped((struct sockaddr *)&client.addr, &client.len);
+
+		mutex_lock(&conn_impl->send_mutex);
+		int reserve_err = conn_poll_tcp_reserve(conn_impl, conn_impl->tcp_conns_count + 1);
```

Comment:
This allows anyone to make the program run out of memory by simply opening TCP connections in a loop.